drivers/rn2xx3/rn2xx3.c

```diff
@@ -0,0 +1,303 @@
+    if (dev->p.pin_reset != GPIO_UNDEF) {
+        gpio_clear(dev->p.pin_reset);
+        xtimer_usleep(RESET_DELAY);
+        gpio_set(dev->p.pin_reset);
+    }
+
+    dev->sleep_timer.callback = _sleep_timer_cb;
+    dev->sleep_timer.arg = dev;
+
+    rn2xx3_sys_set_sleep_duration(dev, RN2XX3_DEFAULT_SLEEP);
+
+    /* sending empty command to clear uart buffer */
+    if (rn2xx3_write_cmd(dev) == RN2XX3_TIMEOUT) {
+        DEBUG("[rn2xx3] init: initialization failed\n");
+        return RN2XX3_TIMEOUT;
+    }
+
+    if (rn2xx3_mac_init(dev) != RN2XX3_OK) {
+        DEBUG("[rn2xx3] mac: initialization failed\n");
+        return RN2XX3_ERR_MAC_INIT;
+    }
+
+    DEBUG("[rn2xx3] init: initialization successful\n");
+    return RN2XX3_OK;
+}
+
+int rn2xx3_sys_reset(rn2xx3_t *dev)
+{
+    memset(dev->cmd_buf, 0, strlen(dev->cmd_buf));
+    snprintf(dev->cmd_buf, 10, "sys reset");
+    int ret = rn2xx3_write_cmd(dev);
+    if (ret == RN2XX3_TIMEOUT || ret == RN2XX3_ERR_SLEEP_MODE) {
+        DEBUG("[rn2xx3] reset: failed\n");
+        return ret;
+    }
+
+    return RN2XX3_OK;
+}
+
+int rn2xx3_sys_sleep(rn2xx3_t *dev)
+{
+    memset(dev->cmd_buf, 0, strlen(dev->cmd_buf));
+    snprintf(dev->cmd_buf, 20, "sys sleep %lu", dev->sleep);
+    if (rn2xx3_write_cmd_no_wait(dev) == RN2XX3_ERR_INVALID_PARAM) {
+        DEBUG("[rn2xx3] sleep: cannot put module in sleep mode\n");
+        return RN2XX3_ERR_INVALID_PARAM;
+    }
+
+    rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_SLEEP);
+    xtimer_set(&dev->sleep_timer, dev->sleep * US_PER_MS);
+
+    return RN2XX3_OK;
+}
+
+int rn2xx3_mac_init(rn2xx3_t *dev)
+{
+    rn2xx3_mac_set_dev_eui(dev, dev->p.loramac.dev_eui);
+    rn2xx3_mac_set_app_eui(dev, dev->p.loramac.app_eui);
+#ifdef LORAMAC_APP_APPKEY
+    rn2xx3_mac_set_app_key(dev, dev->p.loramac.app_key);
+#endif
+#ifdef LORAMAC_APP_SESSION_APPKEY
+    rn2xx3_mac_set_apps_key(dev, dev->p.loramac.app_skey);
+#endif
+#ifdef LORAMAC_NET_SESSION_KEY
+    rn2xx3_mac_set_nwkskey_key(dev, dev->p.loramac.net_skey);
+#endif
+
+    rn2xx3_mac_set_dr(dev, LORAMAC_DEFAULT_DR);
+    rn2xx3_mac_set_tx_power(dev, LORAMAC_DEFAULT_TX_POWER);
+    rn2xx3_mac_set_tx_port(dev, LORAMAC_DEFAULT_TX_PORT);
+    rn2xx3_mac_set_tx_mode(dev, LORAMAC_DEFAULT_TX_MODE);
+    rn2xx3_mac_set_adr(dev, LORAMAC_DEFAULT_ADR);
+    rn2xx3_mac_set_retx(dev, LORAMAC_DEFAULT_RETX);
+    rn2xx3_mac_set_linkchk_interval(dev, LORAMAC_DEFAULT_LINKCHK);
+    rn2xx3_mac_set_rx1_delay(dev, LORAMAC_DEFAULT_RX1_DELAY);
+    rn2xx3_mac_set_ar(dev, LORAMAC_DEFAULT_AR);
+    rn2xx3_mac_set_rx2_dr(dev, LORAMAC_DEFAULT_RX2_DR);
+    rn2xx3_mac_set_rx2_freq(dev, LORAMAC_DEFAULT_RX2_FREQ);
+
+    if (LORAMAC_DEFAULT_JOIN_PROCEDURE == LORAMAC_JOIN_ABP) {
+        DEBUG("[rn2xx3] setting device addr only in ABP mode\n");
+        rn2xx3_mac_set_dev_addr(dev, dev->p.loramac.dev_addr);
+        /* Store parameters to EEPROM */
+        rn2xx3_mac_save(dev);
+    }
+
+    rn2xx3_sys_reset(dev);
```

Comment:
I think so.

Reply:
I finally removed this line: resetting the module was also resetting the previously set keys if not saved to module EEPROM first.
Since only the device address has to be stored to EEPROM before joining the network using ABP, this line was causing problems for other keys.